simplifier/range_simplifier.py

```diff
@@ -317,6 +318,107 @@ def _add_must_value_if_values_are_equal(self, first_const_bound: ConstantBound,
+                self.lower_bound.unsigned = self.lower_bound.signed
+            self.lower_bound.signed = None
+
+    def _combine_singed_lower_unsigned_upper_bounds(self):
+        """
+        Combine signed lower and unsigned upper bound values, i.e. x u<= c_u and x s>= c_s.
+
+        MS = maximum size of a singed value, i.e., for size 4 it is 8
+        1. c_u <= MS and c_s < 0: (x u<= c_u and x s>= c_s) can be simplified to (x u<= c_u)
+        2. c_u <= MS and c_s >= 0: (x u<= c_u and x s>= c_s) can be simplified to (c_s u<= x u<= c_u)
+        3. c_u > MS and c_s >= 0: (x u<= c_u and x s>= c_s) can be simplified to (x s>= c_s)
+        """
+        upper_size_bound_value_signed = eval_constant(self.size_upper_bound.signed, is_signed=True)
+        signed_lower_bound_value = eval_constant(self.lower_bound.signed, is_signed=True)
+        unsigned_upper_bound_value = eval_constant(self.upper_bound.unsigned, is_signed=False)
+        if unsigned_upper_bound_value <= upper_size_bound_value_signed and signed_lower_bound_value < 0:
+            self.lower_bound.signed = None
+        elif unsigned_upper_bound_value <= upper_size_bound_value_signed and signed_lower_bound_value >= 0:
+            if signed_lower_bound_value > 0:
+                self.lower_bound.unsigned = self.lower_bound.signed
+            self.lower_bound.signed = None
+        elif unsigned_upper_bound_value > upper_size_bound_value_signed and signed_lower_bound_value >= 0:
+            self.upper_bound.unsigned = None
+
+    def _combine_singed_upper_unsigned_lower_bounds(self):
+        """
+        Combine signed upper and unsigned lower bound values, i.e. x u>= c_u and x s<= c_s.
+
+        MS = maximum size of a singed value, i.e., for size 4 it is 8
+        1. c_u <= MS and c_s < 0: (x u>= c_u and x s<= c_s) can be simplified to (x s<= c_s)
+        2. c_u > MS and c_s < 0: (x u>= c_u and x s<= c_s) can be simplified to (c_u s<= x s<= c_s)
+        3. c_u > MS and c_s >= 0: (x u>= c_u and x s<= c_s) can be simplified to (x u>= c_u)
+        """
+        upper_size_bound_value_signed = eval_constant(self.size_upper_bound.signed, is_signed=True)
+        signed_upper_bound_value = eval_constant(self.upper_bound.signed, is_signed=True)
+        unsigned_lower_bound_value = eval_constant(self.lower_bound.unsigned, is_signed=False)
+        if unsigned_lower_bound_value <= upper_size_bound_value_signed and signed_upper_bound_value < 0:
+            self.lower_bound.unsigned = None
+        elif unsigned_lower_bound_value > upper_size_bound_value_signed and signed_upper_bound_value < 0:
+            if eval_constant(self.lower_bound.unsigned, is_signed=True) != -upper_size_bound_value_signed - 1:
+                self.lower_bound.signed = self.lower_bound.unsigned
+            self.lower_bound.unsigned = None
+        elif unsigned_lower_bound_value > upper_size_bound_value_signed and signed_upper_bound_value >= 0:
+            self.upper_bound.signed = None
```

Comment:
I find no way to optimize these functions, but maybe we should extract them into a separate class

Reply:
I am not sure whether extracting them to a function makes it better, because we need 2/3 of the class attributes also in these functions and we change them. 
Additionally, it seems strange to have a few simplification strategies in a different class.

Reply:
I split the file into multiple files. Now, we have:
- utils-file containing the functions on top that do not belong to any class
- single_range_simplifier containing the two classes `RichOperand` and `SingleRangeSimplifier`
- expression_values containing the three classes `ConstandBound`, `ExpressionValues` and `BoundRelation`
- current file containing the remaining classes